src/main/java/com/eerussianguy/blazemap/engine/client/LayerRegionTile.java

```diff
@@ -39,7 +43,13 @@ public LayerRegionTile(StorageAccess.Internal storage, BlazeRegistry.Key<Layer>
     public void tryLoad() {
         if(file.exists()) {
             Profilers.FileOps.LAYER_READ_TIME_PROFILER.begin();
-            lock.lockPriority();
+
+            // Trying to greedily acquire the locks with `tryLock()` to minimise delay on game thread
+            // TODO: Figure out a "failure" case so the timed tryLock can be used instead to "give up"
+            // if it's waited too long, thereby not blocking the tick
+            if (!fileLock.readLock().tryLock()) fileLock.readLock().lock();
+            if (!imageLock.writeLock().tryLock()) imageLock.writeLock().lock();
```

Comment:
what's going on here?

Reply:
So, starting with docs ref for `tryLock`: https://docs.oracle.com/en%2Fjava%2Fjavase%2F17%2Fdocs%2Fapi%2F%2F/java.base/java/util/concurrent/locks/ReentrantLock.html#tryLock()

> a call to tryLock() will immediately acquire the lock if it is available, whether or not other threads are currently waiting for the lock. This "barging" behavior can be useful in certain circumstances, even though it breaks fairness.

So, what these two lines are both doing (with their own respective lock) is:
1) Try to acquire the lock with that "barging" behaviour. If the lock happens to be free but several threads want it at that instance, the `tryLock` will _always_ win out. That gives the thread some priority,
2) If the lock isn't free at that instance, `tryLock()` will return false instead of waiting to acquire the lock.
3) Negating that returned value, the if statement will go on to do a normal call to lock the thread that _will_ pause to wait for it.
4) By the end of the line, the thread will have acquired that lock exactly once.

This idiom is being used similar to the way suggested in the function below the one linked above.

Am not sure if the situation will come up much at all that the barging behaviour will actually be triggered, but I figured why not try anyway, given it's on the critical path?

Reply:
Yes, a _single_ RWLock will hold both its read and write locks. However, two _separate_ locks are being acquired here:
- The `fileLock` is getting read locked to protect the LRT file on disk from being changed while it's being read from
- The `imageLock` is getting write protected to protect the representation of the LRT saved to the class object in the in the `image` variable from being read from or written to while it's being populated by the contents being read from file.

Each lock is protecting a different resource. This makes it far more efficient as we're only locking the shared resources we're actually needing to access, and only locking them down as much as we need to protect the integrity of that particular shared resource.

Reply:
This granularity of locking is what most of the commit focusing on locking is about. We always want to lock as little as we possibly can get away with, which means more granular locks and other strategies designed to protect each resource separately. We don't want to unnecessarily block another thread who's working with data that doesn't interact with the data our thread is working on!

Reply:
oh, right, I totally missed that it was two different locks for some reason...